File: src/modules/trees/2_traversals/3_iterative_bfs.ts
class TreeNode {
    data: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(data: number) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class LevelOrderTraversal {
    public levelOrderTraversal(root: TreeNode | null): void {
        if (root === null) return;
        let queue: TreeNode[] = [];
        queue.push(root);
        while (queue.length > 0) {
            let curr: TreeNode = queue.shift()!;
            console.log(curr.data);
            if (curr.left !== null) queue.push(curr.left);
            if (curr.right !== null) queue.push(curr.right);
        }
    }
}

const root = new TreeNode(10); // Create the root node with data 10
root.left = new TreeNode(5); // Create left child with data 5
root.right = new TreeNode(20); // Create right child with data 20
root.left.left = new TreeNode(3); // Create left child of left child with data 3
root.left.right = new TreeNode(8); // Create right child of left child with data 8
root.right.left = new TreeNode(15); // Create left child of right child with data 15
root.right.right = new TreeNode(25); // Create right child of right child with data 25

const traversal = new LevelOrderTraversal();

console.log("Level Order Traversal: ");
traversal.levelOrderTraversal(root); // Output: 10 5 20 3 8 15 25

File: src/modules/trees/2_traversals/3_iterative_bfs.go
package main

import "fmt"

type TreeNode struct {
	data  int
	left  *TreeNode
	right *TreeNode
}

func NewTreeNode(data int) *TreeNode {
	return &TreeNode{data: data}
}

type LevelOrderTraversal struct{}

func (lot *LevelOrderTraversal) breadthFirstSearch(root *TreeNode) {
	if root == nil {
		return
	}

	queue := []*TreeNode{}
	queue = append(queue, root)

	for len(queue) > 0 {
		var curr *TreeNode = queue[0]
		// removing the first element from the queue
		queue = queue[1:]
		fmt.Println(curr.data)
		if curr.left != nil {
			queue = append(queue, curr.left)
		}
		if curr.right != nil {
			queue = append(queue, curr.right)
		}
	}
}

func main() {
	// Create the tree
	root := NewTreeNode(10)
	root.left = NewTreeNode(5)
	root.right = NewTreeNode(20)
	root.left.left = NewTreeNode(3)
	root.left.right = NewTreeNode(6)
	root.right.left = NewTreeNode(16)
	root.right.right = NewTreeNode(61)

	traversal := &LevelOrderTraversal{}

	fmt.Println("Level Order Traversal: ")
	traversal.breadthFirstSearch(root)
}
